lambda/trmc.ml

```diff
@@ -351,27 +365,31 @@ let rec choice ctx t =
             direct = (fun () -> Lapply apply);
           }
       | _nontail -> raise No_trmc
-    with No_trmc -> Return (Lapply apply)
+    with No_trmc -> Choice.Return (Lapply apply)
 
   and choice_makeblock ctx (tag, flag, shape) blockargs loc =
     let k new_flag new_block_args =
       Lprim (Pmakeblock (tag, new_flag, shape), new_block_args, loc) in
     let choices = List.map (choice ctx) blockargs in
-    match find_settable choices with
-    | Error all_returns -> Return (k flag all_returns)
+    match Choice.find_settable choices with
+    | Error all_returns ->
+        let all_returns = traverse_list ctx all_returns in
+        Choice.Return (k flag all_returns)
     | Ok { rev_returns; settable; tail_choices } ->
         begin
           (* fail if this settable position is not unique *)
-          match find_settable tail_choices with
+          match Choice.find_settable tail_choices with
           | Error _all_returns -> ()
           | Ok _another_settable ->
               failwith "TODO proper error/warning: ambiguous settable position"
         end;
+        let rev_returns = traverse_list ctx rev_returns in
+        let tail_terms = traverse_list ctx (List.map return tail_choices) in
```

Comment:
Why are we now traversing the returns and remaining choices?  Isn't this redundant on values returned by `choice`?

Reply:
I believe that you are correct, this is a mistake that results in traversing the terms twice. Good catch! I will try to see if I can observe the difference by testing, and fix this.

Reply:
So yes, we can observe double-traversal because the TRMC transformation is applied twice. (I was curious how the result looks like: 4 versions are generated, the code is correct, but it's a bit disappointing because the twice-trmc function is not called by the other versions, given that in a once-TRMCed function there are no recursive call remaining in TRMC-but-not-tailcall position.)